File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
class Solution(object):
    def check(self, s, word_count, word_length):
        # Count occurrences of words in the current substring
        current_word_count = Counter()
        for i in range(0, len(s), word_length):
            word = s[i : i + word_length]
            current_word_count[word] += 1
        
        return current_word_count == word_count

    def findSubstring(self, s, words):
        result = []

        if s == "".join(words):
            return [0]

        word_count = Counter(words)
        word_length = len(words[0])
        total_length = len(words) * word_length

        for i in range(len(s) - total_length + 1):
            substring = s[i : i + total_length]
            if self.check(substring, word_count, word_length):
                result.append(i)

        return result

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> result;
        if (s.empty() || words.empty()) return result;
        
        unordered_map<string, int> wordCount;
        for (const string& word : words) {
            wordCount[word]++;
        }
        
        int len = s.length();
        int wordLen = words[0].length();
        int wordsNum = words.size();
        int totalLen = wordLen * wordsNum;
        
        for (int offset = 0; offset < wordLen; offset++) {
            unordered_map<string, int> windowCount;
            int matched = 0;
            
            for (int i = offset; i <= len - wordLen; i += wordLen) {
                string word = s.substr(i, wordLen);
                
                windowCount[word]++;
                matched++;
                
                if (matched > wordsNum) {
                    string leftWord = s.substr(i - totalLen, wordLen);
                    windowCount[leftWord]--;
                    if (windowCount[leftWord] == 0) {
                        windowCount.erase(leftWord);
                    }
                    matched--;
                }
                
                if (matched == wordsNum && windowCount == wordCount) {
                    result.push_back(i - totalLen + wordLen);
                }
            }
        }
        
        return result;
    }
};